ExampleMod/Content/NPCs/FallenSoul.cs

```diff
@@ -0,0 +1,137 @@
+
+		// Allows us to do something when the NPC is hit. Here, we spawn dust.
+		public override void HitEffect(int hitDirection, double damage) {
+			if (npc.life > 0) {
+				for (int i = 0; i < damage / npc.lifeMax * 100; i++) {
+					Dust dust = Dust.NewDustDirect(npc.position, npc.width, npc.height, 192, hitDirection, -1f, 100, new Color(100, 100, 100, 100), 1f);
+					dust.noGravity = true;
+				}
+
+				return;
+			}
+
+			for (int i = 0; i < 50; i++) {
+				Dust dust = Dust.NewDustDirect(npc.position, npc.width, npc.height, 192, 2 * hitDirection, -2f, 100, new Color(100, 100, 100, 100), 1f);
+				dust.noGravity = true;
+			}
+		}
+
+		// Allows the NPC to talk with the player, even if it isn't a town NPC.
+		public override bool CanChat() => true;
+
+		// Allows us to choose what the NPC says when talked to.
+		public override string GetChat() {
+			// npc.SpawnedFromStatue value is kept when the NPC is transformed.
+			switch (Main.rand.Next(npc.SpawnedFromStatue ? 5 : 3)) {
+				case 0:
+					return "Thank you, now I don't have to haunt random people anymore, only you.";
+				case 1:
+					return "Keep breaking those evil altars, me and many others were cursed to haunt anyone who did so.";
+				case 2:
+					return "Can you help me get into heaven?";
+				default:
+					return "Please stop messing with that haunted statue. Don't you know what \"RIP\" means?";
+			}
+		}
+
+		// Allows us to set when our chat buttons will say.
+		public override void SetChatButtons(ref string button, ref string button2) => button = "Send to heaven";
+
+		// Allows us to do stuff depending on which button was pressed.
+		public override void OnChatButtonClicked(bool firstButton, ref bool shop) {
+			if (firstButton) {
+				// Hit the NPC for about 500 damage if the first button is clicked.
+				Main.LocalPlayer.ApplyDamageToNPC(npc, Main.DamageVar(500), 5f, Main.LocalPlayer.direction, true);
+			}
+		}
+
+		// Only draw the NPC's health bar when the player is at a certain distance or closer.
+		public override bool? DrawHealthBar(byte hbPosition, ref float scale, ref Vector2 position) {
+			float distance = npc.Distance(Main.player[npc.target].Center);
+
+			if (distance <= 200) {
+				if (distance > 100) {
+					// Make the health bar become smaller the farther away the NPC is.
+					scale *= (100 - (distance - 100)) / 100;
+				}
+
+				return null;
+			}
+
+			return false;
+		}
+
+		// Make the NPC invisible when far away from the player.
+		// CustomBehavior is a custom hook in HoverNPC.cs.
+		public override void CustomBehavior(ref float ai) {
+			float distance = npc.Distance(Main.player[npc.target].Center);
+
+			if (distance <= 250) {
+				npc.alpha = 100;
+
+				if (distance > 100) {
+					// Make the NPC fade out the farther away the NPC is.
+					npc.alpha += (int)(155 * ((distance - 100) / 150));
+				}
+
+				return;
+			}
+
+			npc.alpha = 255;
```

Comment:
same as above, just use `100 + Utils.SmoothStep((250, 100, distance) * 155`

Seriously, maybe tML should have a `Utils.Lerp(float x1, float y1, float x2, float y2, float x)`

Also yeah, lets not use `Utils.SmoothStep` because it's definition is really confusing when there's also `MathHelper.SmoothStep` which does something completely different